src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs

```diff
@@ -0,0 +1,308 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace IOperationGenerator
+{
+    internal enum NullHandling
+    {
+        Allow,
+        Disallow,
+        Always,
+        NotApplicable // for value types
+    }
+
+    internal sealed class IOperationClassWriter
+    {
+        private TextWriter _writer = null!;
+        private readonly string _location;
+        private readonly Tree _tree;
+
+        private IOperationClassWriter(Tree tree, string location)
+        {
+            _tree = tree;
+            _location = location;
+        }
+
+        public static void Write(Tree tree, string location)
+        {
+            new IOperationClassWriter(tree, location).WriteFiles();
+        }
+
+        private int _indent;
+        private bool _needsIndent = true;
+
+        private void Write(string format)
+        {
+            if (_needsIndent)
+            {
+                _writer.Write(new string(' ', _indent * 4));
+                _needsIndent = false;
+            }
+            _writer.Write(format);
+        }
+
+        private void WriteLine(string format)
+        {
+            Write(format);
+            _writer.WriteLine();
+            _needsIndent = true;
+        }
+
+        private void Blank()
+        {
+            _writer.WriteLine();
+            _needsIndent = true;
+        }
+
+        private void Brace()
+        {
+            WriteLine("{");
+            Indent();
+        }
+
+        private void Unbrace()
+        {
+            Outdent();
+            WriteLine("}");
+        }
+
+        private void Indent()
+        {
+            ++_indent;
+        }
+
+        private void Outdent()
+        {
+            --_indent;
+        }
+
+        private void WriteFiles()
+        {
+            foreach (var grouping in _tree.Types.OfType<AbstractNode>().GroupBy(n => n.Namespace))
+            {
+                var @namespace = grouping.Key ?? "Operations";
+                var outFileName = Path.Combine(_location, $"{@namespace}.Generated.cs");
+                using (_writer = new StreamWriter(File.Open(outFileName, FileMode.Create), Encoding.UTF8))
+                {
+                    writeHeader();
+                    WriteUsing("System");
+                    WriteUsing("System.Collections.Immutable");
+
+                    if (grouping.Key is object)
```

Comment:
>if (grouping.Key is object) [](start = 20, length = 27)

I guess it doesn't matter for an internal tool but will this include an unnecessary using if the XML contains an explicit "Operations" namespace?

Reply:
Yes, we would. I could add assert that it doesn't.